### ISSUE TITLE
Does not work with Elixir 1.15

### ISSUE THREAD

**Issue by eKIK** [bryanjos/geo_postgis#161]
I have an project making use of geo_postgis and it works fine using Elixir 1.14.
When changing version to 1.15 the compilation fails:

```
** (ArgumentError) unknown type Geo.PostGIS.Geometry for field :boundary
    (ecto 3.10.3) lib/ecto/schema.ex:2318: Ecto.Schema.check_field_type!/4
    (ecto 3.10.3) lib/ecto/schema.ex:1931: Ecto.Schema.__field__/4
    lib/ls/gis/schema/municipality_schema.ex:11: (module)
```

Running `mix deps.clean && mix.deps.get` and such doesn't seem to help out.
It's not always 100% easy to reproduce. I guess it's because you might have your files compiled with 1.14 lying around, which makes it work.

I believe this could be related to the code pruning introduced in Elixir 1.15 (https://hexdocs.pm/elixir/1.15/changelog.html)

> Due to the code path pruning, if you have an application or dependency that does not specify its dependencies on Erlang and Elixir application, it may no longer compile successfully in Elixir v1.15. 

Since geo_postgis doesn't have a hard dependency on ecto (`{:ecto_sql, "~> 3.0", optional: true, only: :test}`) I would expect the code pruning to kick in. This matches what I'm seeing because in `geometry.ex` the code `if Code.ensure_loaded?(Ecto.Type) do` always seems to return `false` for me (using Elixir 1.15).

Setting `prune_code_paths: false` for the project in `mix.exs` does fix the problem for me, which also indicates that this is the issue.

**Comment by jakespracher** [comment]
Believe I just encountered this same issue. Following

**Comment by theolampert** [comment]
I worked around this for the moment [here](https://github.com/bryanjos/geo_postgis/compare/master...theolampert:geo_postgis:master) by just explicitly setting the Ecto version I was using and removing the conditional check, I'm not experienced enough with Elixir to know what the proper fix would be.

**Comment by eKIK** [comment]
My very hacky workaround (for now) was to just simply copy the contents of the geo_postgis "lib" folder into my own "lib/vendor/geo_postgis" folder (which I created). This way it's part of the applications library files, and won't get pruned by the compiler. @theolampert 's approach is neater in a way, but I was just too lazy to change the code and put it in a repo myself :).

I guess one option would be to extract the Ecto specific parts from the Postgrex ones, and have that as a separate library (geo_postgis_ecto). That way we could even specialize the docs for Ecto etc. It'd make it easier (imo) for people who wanted to use Ecto to interact with Postgis to get started. Obviously this would be a big change, and dealing with backwards compatibility would be a hassle, so I'm not sure if it's worth it. Potentially there's some clever thing one could do...but I don't know enough about Elixir either...

**Comment by iloveitaly** [comment]
Another option here is [`mix compile --force`](https://elixirforum.com/t/vegalite-viewer-not-working/57254/7) which fixed the issue for me.

**Comment by josevalim** [comment]
What is your precise Elixir version? There was a bug in Elixir v1.15.x but it should be addressed on v1.15.4. If not, please provide a way to reproduce the issue and I will take a look at it. But from a first glance, I don't see anything wrong in geo_postgis.

**Comment by eKIK** [comment]
@josevalim Spot on! After upgrading to 1.15.4 it all worked great.

I just had to `mix clean && mix deps.clean` but that's most likely because of the changes I had done pulling in geo_postgis into my own lib folder.

I'm closing this issue since it wasn't an issue with geo_postgis in the end. Just make sure you're on Elixir 1.15.4+ and it should work just fine.
